### PR TITLE
Fix for invalid backend type 

### LINKED ISSUE #1107: [Backend Service] Backend Registration is Successful Even for a Non-Existence Storage Backend Type
**Issue/Feature Description:** Backend Registration is Successful even for a Non-Existence Storage Backend Type
This should be 400 Error message for this operation

**Why this issue to fixed / feature is needed(give scenarios or use cases):** To provide Negative Scenarios for Backend Creation

**How to reproduce, in case of a bug:**

Create backend for the following:
POST http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/backends

```
{
  "name": "gcp-backend-file",
  "type": "gcp-file-fake",
  "Region": "region",
  "Access": "access_key",
  "Security": "secret_key"
}
```

**Other Notes /  Environment Information: (Please give the env information, log link or any useful information for this issue)**

### PR DESCRIPTION
**Pull request by vineela1999**
**What type of PR is this?**
> /kind bug fix

**What this PR does / why we need it**:
when a create backend request is sent, type of the backend is to be verified to proceed.
**Which issue(s) this PR fixes**:

Fixes #1107 

**Test Report Added?**:
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
> /kind TESTED
> /kind NOT-TESTED

**Test Report**:
 when an invalid backend type is given:
image
image
When a valid backend type is given:
Screenshot from 2020-11-17 18-42-34
Screenshot from 2020-11-17 18-44-23

**Special notes for your reviewer**:

### CODE REVIEW

File: api/pkg/backend/service.go [hunk 1]
```diff
@@ -205,17 +204,35 @@ func (s *APIService) CreateBackend(request *restful.Request, response *restful.R
...
+	_, foundType := typeExists(storageTypes.Types, backendDetail.Type)
+	if !foundType {
+		log.Errorf("Failed to retrieve backend type: %v\n", err)
+		response.WriteError(http.StatusBadRequest, err)
+		return
+	}
 	res, err := s.backendClient.CreateBackend(ctx, &backend.CreateBackendRequest{Backend: backendDetail})
 	if err != nil {
-		log.Errorf("failed to create backend: %v\n", err)
+		log.Errorf("Failed to create backend: %v\n", err)
```
Comment (himanshuvar) [inline api/pkg/backend/service.go:218]: Same as above

File: api/pkg/backend/service.go [hunk 2]
```diff
@@ -205,17 +204,35 @@ func (s *APIService) CreateBackend(request *restful.Request, response *restful.R
...
 		response.WriteError(http.StatusInternalServerError, err)
 		return
 	}
 
-	log.Info("Create backend successfully.")
+	log.Info("Created backend successfully.")
 	response.WriteEntity(res.Backend)
 }
 
+func typeExists(slice []*backend.TypeDetail, inputType string) (int, bool) {
```
Comment (himanshuvar) [inline api/pkg/backend/service.go:227]: Do we need to return two parameters, Are we using index?
Only boolean is not sufficient?

File: api/pkg/backend/service.go [hunk 3]
```diff
@@ -289,48 +306,14 @@ func (s *APIService) ListType(request *restful.Request, response *restful.Respon
...
-	if err != nil {
-		log.Errorf("failed to list types: %v\n", err)
-		response.WriteError(http.StatusInternalServerError, err)
-		return
-	}
-
+	storageTypes,_:=s.listStorageType(ctx,request,response)
 	log.Info("List types successfully.")
-	response.WriteEntity(res)
+	response.WriteEntity(storageTypes)
```
Comment (himanshuvar) [inline api/pkg/backend/service.go:313]: Do we need to handle Exception/Error before writing to response

File: api/pkg/backend/service.go [hunk 4]
```diff
@@ -205,17 +204,35 @@ func (s *APIService) CreateBackend(request *restful.Request, response *restful.R
 	actx := request.Attribute(c.KContext).(*c.Context)
 	backendDetail.TenantId = actx.TenantId
 	backendDetail.UserId = actx.UserId
+
+	storageTypes,_:= s.listStorageType(ctx,request,response)
```
Comment (himanshuvar) [inline api/pkg/backend/service.go:208]: Please handle Exception/Error, there might be a nil pointer

**Comment by vineela1999** [comment]
@himanshuvar code rework done

File: api/pkg/backend/service.go [hunk 5]
```diff
@@ -205,32 +205,36 @@ func (s *APIService) CreateBackend(request *restful.Request, response *restful.R
 	backendDetail.TenantId = actx.TenantId
 	backendDetail.UserId = actx.UserId
 
-	storageTypes,_:= s.listStorageType(ctx,request,response)
-
-	_, foundType := typeExists(storageTypes.Types, backendDetail.Type)
+	storageTypes,err:= s.listStorageType(ctx,request,response)
+	if err != nil {
+		log.Errorf("failed to create backend: %v\n", err)
```
Comment (himanshuvar) [inline api/pkg/backend/service.go:210]: failed to list backend?

File: api/pkg/backend/service.go [hunk 6]
```diff
@@ -205,32 +205,36 @@ func (s *APIService) CreateBackend(request *restful.Request, response *restful.R
...
 	res, err := s.backendClient.CreateBackend(ctx, &backend.CreateBackendRequest{Backend: backendDetail})
 	if err != nil {
-		log.Errorf("Failed to create backend: %v\n", err)
+		log.Errorf("failed to create backend: %v\n", err)
 		response.WriteError(http.StatusInternalServerError, err)
 		return
 	}
 
-	log.Info("Created backend successfully.")
+	log.Info("Create backend successfully.")
```
Comment (himanshuvar) [inline api/pkg/backend/service.go]: Created was appropriate

**Review by himanshuvar — CHANGES_REQUESTED** [review]
Minor Typo comments

File: api/pkg/backend/service.go [hunk 7]
```diff
@@ -216,6 +228,15 @@ func (s *APIService) CreateBackend(request *restful.Request, response *restful.R
 	response.WriteEntity(res.Backend)
 }
 
+func typeExists(slice []*backend.TypeDetail, inputType string) (bool) {
+	for _, item := range slice {
+		if item.Name == inputType {
```
Comment (kumarashit) [inline api/pkg/backend/service.go:233]: Put a debug level log message before returning

**Comment by vineela1999** [comment]
@himanshuvar @kumarashit 
changes done

**Review by kumarashit — APPROVED** [review]
LGTM

File: api/pkg/backend/service.go [hunk 8]
```diff
@@ -205,17 +204,40 @@ func (s *APIService) CreateBackend(request *restful.Request, response *restful.R
 	actx := request.Attribute(c.KContext).(*c.Context)
 	backendDetail.TenantId = actx.TenantId
 	backendDetail.UserId = actx.UserId
+
+	storageTypes,err:= s.listStorageType(ctx,request,response)
```
Comment (wisererik) [inline api/pkg/backend/service.go:208]: need space between , and err

File: api/pkg/backend/service.go [hunk 9]
```diff
@@ -289,48 +311,19 @@ func (s *APIService) ListType(request *restful.Request, response *restful.Respon
...
-		log.Errorf("get filter failed: %v\n", err)
-		response.WriteError(http.StatusInternalServerError, err)
-		return
-	}
-	listTypeRequest.Filter = filter
-
+	log.Info("Received request for backends type list.")
 	ctx := context.Background()
-	res, err := s.backendClient.ListType(ctx, listTypeRequest)
+	storageTypes,err:=s.listStorageType(ctx,request,response)
```
Comment (wisererik) [inline api/pkg/backend/service.go:316]: storageTypes,err:=s.listStorageType(ctx,request,response) -> storageTypes, err := s.listStorageType(ctx, request, response)

**Review by wisererik — CHANGES_REQUESTED** [review]
it's better to format code

**Comment by vineela1999** [comment]
@wisererik done with change

**Review by wisererik — APPROVED** [review]

**Pull request merged**